optimum/quanto/library/extensions/extension.py

```diff
@@ -24,6 +23,9 @@ def __init__(
         self.build_directory = os.path.join(root_dir, "build")
         self._lib = None
 
+        # There is no reason not to build ahead of runtime.
```

Comment:
It does relate as otherwise
```
from optimum.quanto.library.extensions.cuda import ext
```
does not load the kernel to be usable with `torch.ops.quanto_ext.fp8_marlin_gemm`.

Reply:
The extensions are JIT, and the current design allows to only build them:
- if they can be compiled,
- when they are required.
You are breaking both paradigms by:
- directly importing a symbol that is not defined on non CUDA platforms,
- forcing a compilation that will fail.

Reply:
Yes I get that. Maybe I need a
```python
if is_cuda_available():
    from optimum.quanto.library.extensions.cuda import ext
    _ = ext.lib
```

Off topic here: what is the issue with compiling ahead of time, depending on the plateform? Most libraries do that (pytorch, vllm, tgi, autoawq, etc.) and ship wheels depending on the plateform, actually most programs that I know of do that. This avoids issues where you want to use a feature, and have to wait tens of seconds each time before your code does run.

What I am saying is that you need something along these lines to be able to use `TORCH_LIBRARY`. I did not manage to have JIT TORCH_LIBRARY working (explicitly calling e.g. `_ = ext.lib` in forward may or may not be okay)

Reply:
Maintaining wheels for each platform is very expensive because of the combinations (for CUDA at least 2, maybe 3 versions are required for Windows AND Linux), not even mentioning the fact that for linux platforms you need to create manylinux wheels, which most libraries pretend to do but actually don't.
Deploying a source package with JIT extensions is much simpler, and IMHO more effective, because you basically support at once all combinations pytorch supports.
I just strictly followed the pytorch guidelines.

Reply:
Anyway, some of quanto design choices have precisely made to address shortcomings of "the other libraries" (no finger-pointing but the UX of some of them is just terrible), so these are not really subject to discussion:

- dispatch happens at the Tensor level, to be able to switch between devices (as opposed to using custom modeling that works for CUDA only),
- extensions are JIT (as opposed to having to manually tweak your setup to match what the binary extension expects),
- quantized tensors are differentiable to support QAT,
- quantized models can be serialized and reloaded using a device/kernel agnostic format (as opposed to using a custom packing every time and having incompatible weights).